File: src/main/java/com/github/davinkevin/transmissionrss/batch/synchronization/items/FeedProcessor.java
package com.github.davinkevin.transmissionrss.batch.synchronization.items;

import com.github.davinkevin.transmissionrss.batch.synchronization.items.model.RssItemsWithFeedUrl;
import com.github.davinkevin.transmissionrss.rss.service.RssTorrentService;
import io.vavr.collection.List;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.batch.item.ItemProcessor;
import org.springframework.stereotype.Component;

import java.net.URL;

/**
 * Created by kevin on 08/04/2018
 */
@Slf4j
@Component
@RequiredArgsConstructor
public class FeedProcessor implements ItemProcessor<URL, List<RssItemsWithFeedUrl>> {

    private final RssTorrentService rssTorrentService;

    @Override
    public List<RssItemsWithFeedUrl> process(URL url) {
        return rssTorrentService
                .parse(url)
                .map(d -> d.getRootElement().getChild("channel").getChildren("item"))
                .map(List::ofAll)
                .getOrElse(List::empty)
                .map(e -> RssItemsWithFeedUrl.from(e, url));
    }
}

File: src/main/kotlin/com/github/davinkevin/transmissionrss/batch/synchronization/feeds/items/FeedProcessor.kt
package com.github.davinkevin.transmissionrss.batch.synchronization.feeds.items

import arrow.core.getOrElse
import com.github.davinkevin.transmissionrss.rss.RssTorrentService
import org.springframework.batch.item.ItemProcessor
import org.springframework.stereotype.Component
import java.net.URL
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter.RFC_1123_DATE_TIME

@Component
class FeedProcessor(val rssTorrentService: RssTorrentService): ItemProcessor<URL, FeedWithItems> {

    override fun process(url: URL?): FeedWithItems {
        val items = rssTorrentService
                .parse(url!!)
                .map { it.rootElement.getChild("channel").getChildren("item") }
                .getOrElse { ArrayList() }
                .map {
                    ItemsWithStatus(
                            title = it.getChild("title").textTrim,
                            link = URL(it.getChild("link").textTrim),
                            guid = it.getChild("guid").textTrim,
                            pubDate = ZonedDateTime.parse(it.getChild("pubDate").textTrim, RFC_1123_DATE_TIME)
                    )
                }
        return FeedWithItems(url, items)
    }

}